### ISSUE TITLE
Sparse dense matmul with variable values

### ISSUE THREAD

**Issue by jocampo2** [tensorflow/addons#2772]
**Describe the feature and the current behavior/state.**
Currently tensorflow doesn't support a sparse dense matmul with variables as the values of the sparse tensor (Though there are some hacks). i.e **y**=W**x** where W is a sparse matrix and is learnable

I have an implementation see this [google colab file](https://colab.research.google.com/drive/1u4mQiVi91W16HcPfTLucX9DYHV4qy_Mn#scrollTo=QQevJayeSogN), using either autodiff or a custom gradient(which is more efficient though needs a precompute), and is tested against autodiff. Also works with GPU. Can we add the SparseLayer with custom gradients as a feature in tensorflow_addons?

- Are you willing to contribute it: yes
- Are you willing to maintain it going forward?: yes
- Is there a relevant academic paper?: https://arxiv.org/abs/2209.13603 (In the appendix)
- Does the relavent academic paper exceed 50 citations?: not yet, it was put up last month

**Which API type would this fall under (layer, metric, optimizer, etc.)**
layer

**Who will benefit with this feature?**
People have been asking for this on github and stackoverflow frequently. This is a sparse matrix multiplication layer which is widely used, having learnable values should prove useful for many.

**Comment by seanpmorgan** [comment]
TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
